x-pack/platform/plugins/shared/streams/server/lib/streams/assets/query/query_client.ts

```diff
@@ -0,0 +1,342 @@
+    const currentIds = new Set(currentQueryLinks.map((link) => link.query.id));
+    const nextIds = new Set(queries.map((query) => query.id));
+
+    const nextQueriesToCreate = queries
+      .filter((query) => !currentIds.has(query.id))
+      .map((query) => toQueryLink(query, stream));
+
+    const currentQueriesToDelete = currentQueryLinks.filter((link) => !nextIds.has(link.query.id));
+
+    const currentQueriesToDeleteBeforeUpdate = currentQueryLinks.filter((link) =>
+      queries.some((query) => query.id === link.query.id && hasBreakingChange(link.query, query))
+    );
+
+    const nextQueriesUpdatedWithBreakingChange = queries
+      .filter((query) =>
+        currentQueryLinks.some(
+          (link) => link.query.id === query.id && hasBreakingChange(link.query, query)
+        )
+      )
+      .map((query) => toQueryLink(query, stream));
+
+    const nextQueriesUpdatedWithoutBreakingChange = queries
+      .filter((query) =>
+        currentQueryLinks.some(
+          (link) => link.query.id === query.id && !hasBreakingChange(link.query, query)
+        )
+      )
+      .map((query) => toQueryLink(query, stream));
+
+    await this.uninstallQueries([...currentQueriesToDelete, ...currentQueriesToDeleteBeforeUpdate]);
+    await this.installQueries(
+      [...nextQueriesToCreate, ...nextQueriesUpdatedWithBreakingChange],
+      nextQueriesUpdatedWithoutBreakingChange,
+      stream
+    );
+
+    await this.dependencies.assetClient.syncAssetList(
+      stream,
+      queries.map((query) => ({
+        [ASSET_ID]: query.id,
+        [ASSET_TYPE]: 'query',
+        query,
+      })),
+      'query'
+    );
+  }
+
+  public async upsert(stream: string, query: StreamQuery) {
+    const { assetClient } = this.dependencies;
+
+    if (!this.isSignificantEventsEnabled) {
+      return await assetClient.linkAsset(stream, {
+        [ASSET_ID]: query.id,
+        [ASSET_TYPE]: 'query',
+        query,
+      });
+    }
+
+    const currentQueryLinks = await this.dependencies.assetClient.getAssetLinks(stream, ['query']);
+    const nextQueries = [
+      ...currentQueryLinks.filter((link) => link.query.id !== query.id),
+      toQueryLink(query, stream),
+    ];
+
+    await this.syncQueries(
+      stream,
+      nextQueries.map((link) => link.query)
+    );
+  }
+
+  public async delete(stream: string, queryId: string) {
+    const { assetClient } = this.dependencies;
+
+    if (!this.isSignificantEventsEnabled) {
+      return await assetClient.unlinkAsset(stream, {
+        [ASSET_TYPE]: 'query',
+        [ASSET_ID]: queryId,
+      });
+    }
+
+    const currentQueryLinks = await this.dependencies.assetClient.getAssetLinks(stream, ['query']);
+    const nextQueries = currentQueryLinks.filter((link) => link.query.id !== queryId);
+    await this.syncQueries(
+      stream,
+      nextQueries.map((link) => link.query)
+    );
+  }
+
+  public async bulkOperations(
```

Comment:
should we just call this `bulk`? bit of a nitpick of course.

Reply:
should we just make internal methods call `bulk`? would that help in terms of repetitiveness?

Reply:
I could change upsert and delete to call bulk which then will call sync 😅